### PR TITLE
add parameter to specify a Xena module number

### PR DESCRIPTION
**Pull request by grdumas**
Adds argument parser option, allows user to specify a particular module of a Xena chassis.

(@ctrautma for reference)

### CODE REVIEW

**Comment by grdumas** [comment]
Revised commit; adds Xena throughput test functionality.

**Comment by k-rister** [comment]
@ctrautma I think we can restart work on this.  Obviously the first step we need to do is fix the merge conflicts.

**Comment by ctrautma** [comment]
OK I will take a look. 

----- Original Message -----

**Review by k-rister — CHANGES_REQUESTED** [review]

File: README-xena-trafficgen-integration.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+5. Install TRex
+   * Within trafficgen, open `install-trex.sh` for editing
+   * Make the following change:
+        * original: `if curl --output ${tarfile} ${trex_url} && tar zxf ${tarfile}; then`
+        * modified: ``if curl --insecure --output ${tarfile} ${trex_url} && tar zxf ${tarfile}; then``
+   * As root or using `sudo`, run: `./install-trex.sh`
+   * After install completes, replace the following files with copies from test ZIP `fix_trex_install` (or make changes listed in Known Issues):
+        * `trex_client.py` --> `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/trex_client.py`
+        * `trex_conn.py` --> `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/trex_conn.py`
+        * `trex_global_stats.py` --> `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/stats/trex_global_stats.py`
```
[the remark this answers is not in the document]
Comment (grdumas) [reply README-xena-trafficgen-integration.md:48]: The TRex inclusions are because of the following line from the beginning of binary-search.py:
`from trex_tg_lib import *`

Importing everything from trex_tg_lib in turn kicks off a chain of other imports - end result is that without TRex installed on a system, binary search crashes. For the sake of clarity I'll address the other questions in their own respective comments.

[the remark this answers is not in the document]
Comment (k-rister) [reply README-xena-trafficgen-integration.md:48]: Ok, that makes sense.  I'll have to think if there is a way to work around this.  Having to install TRex when you aren't using it just seems so broken...

[the remark this answers is not in the document]
Comment (k-rister) [reply README-xena-trafficgen-integration.md:48]: @ctrautma

I put together a patch that I think will solve these TRex dependency issues:

https://github.com/k-rister/trafficgen/commit/fa7235c1853061fa3e20c5f8ee7d6c1ac48d8e23

Can you have somebody test if this patch allows the Valkrie2544 stuff (this PR) to work without having the TRex stuff present as is described in the proposed README files?

File: README-xena-trafficgen-integration.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  * Change line 110 from: `self.async.set_as_zombie()` to: `self.async_.set_as_zombie()`
+  * Change line 143 from: `return ( self.async.last_data_recv_ts is not None and ((time.time() - self.async.last_data_recv_ts) <= 3) )` to `return ( self.async_.last_data_recv_ts is not None and ((time.time() - self.async_.last_data_recv_ts) <= 3) )`
+  * Change line 189 from: `rc = self.async.connect()` to `rc = self.async_.connect()`
+
+* SyntaxError from `trex_global_stats.py`
+  * Open `/opt/trex/current/automation/trex_control_plane/interactive/trex/common/stats/trex_global_stats.py`
+  * Change line 81 from: `("async_util.", "{0}% / {1}".format( format_threshold(round_float(self.client.conn.async.monitor.get_cpu_util()), [85, 100], [0, 85]),` to: `("async_util.", "{0}% / {1}".format( format_threshold(round_float(self.client.conn.async_.monitor.get_cpu_util()), [85, 100], [0, 85]),`
+  * Change line 82 from: `format_num(self.client.conn.async.monitor.get_bps() / 8.0, suffix = "B/sec"))),` to: `format_num(self.client.conn.async_.monitor.get_bps() / 8.0, suffix = "B/sec"))),`
```
[the remark this answers is not in the document]
Comment (grdumas) [reply README-xena-trafficgen-integration.md:138]: Egads, I thought I had commented on this already - my apologies! Back at the end of August I wrote this README as a "setup from scratch" sort of guide, which presumed the user had nothing already installed or configured. I walked through the process using a fresh Fedora 32 install, and wrote down the bare minimum steps necessary just to get the Xena functionality working. 

(At the time of writing the patch) binary-search.py required that a user also install TRex in order to function (as mentioned in a previous comment). But even after installing TRex I still couldn't run binary-search.py due to several crash-inducing errors I'd never seen before. I still don't know why these problems suddenly popped up, but I eventually traced the errors back to the TRex library. Perhaps some update from Cisco unintentionally introduced a bug? I also can't rule out that the problem might have been self-inflicted from some flaw in my install/setup process.

Whatever the cause, I found that binary-search.py would stop crashing (at least enough so that the Xena bits worked) if I made a few tweaks to the TRex install. Hacky? Absolutely. But I lacked the necessary knowledge or skills to sort this problem out (or the time to develop such competencies). So I wrote those workarounds into the README in the hopes they'd be "good-enough" for folks needing to use Xena in the short term, at least while the TRex dependency issues  persisted.

TLDR; the TRex changes were just duct tape and will hopefully become unneeded/irrelevant/obsolete and removed as soon as possible.
